File: net/JetBrains.Serialization/src/FileInfos/MachO/MachoFileMetaInfo.cs
using JetBrains.FormatRipper.Dmg;
using JetBrains.FormatRipper.MachO;
using Newtonsoft.Json;

namespace JetBrains.Serialization.FileInfos.MachO;

[JsonObject(MemberSerialization.OptIn)]
public class MachoFileMetaInfo : IFileMetaInfo
{
  [JsonProperty("Metadata")] public MachoFileMetadata Metadata;

  [JsonConstructor]
  public MachoFileMetaInfo(MachoFileMetadata metadata)
  {
    Metadata = metadata;
  }

  public void ModifyFile(Stream stream, byte[] signature)
  {
    if (Metadata.FileSize > stream.Length)
    {
      stream.SetLength(Metadata.FileSize);
    }

    stream.Position = Metadata.MachoOffset;

    var headerBytes = Metadata.HeaderMetainfo.ToByteArray(Metadata.IsBe);

    stream.Write(headerBytes, 0, headerBytes.Length);

    foreach (var metadataLoadCommand in Metadata.LoadCommands)
    {
      stream.Position = metadataLoadCommand.Offset + Metadata.MachoOffset;
      var commandBytes = metadataLoadCommand.ToByteArray();
      stream.Write(commandBytes, 0, commandBytes.Length);
    }

    stream.Position = Metadata.CodeSignatureInfo.SuperBlobStart + Metadata.MachoOffset;
    var codeSignatureInfoBytes = Metadata.CodeSignatureInfo.ToByteArray();
    stream.Write(codeSignatureInfoBytes, 0, codeSignatureInfoBytes.Length);

    foreach (var blob in Metadata.CodeSignatureInfo.Blobs)
    {
      stream.Position = Metadata.CodeSignatureInfo.SuperBlobStart + Metadata.MachoOffset + blob.Offset;
      if (blob.Magic == CSMAGIC_CONSTS.CMS_SIGNATURE)
      {
        blob.Length = signature.Length + 8;
        blob.Content = signature;
      }

      if (blob.Type == (uint)CSMAGIC_CONSTS.CODEDIRECTORY)
      {
        stream.Write(blob.Content, 0, blob.Content.Length);
      }
      else
      {
        var blobBytes = blob.ToByteArray();
        stream.Write(blobBytes, 0, blobBytes.Length);
      }
    }
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/fileInfos/MachO/MachoFileMetaInfo.kt
package com.jetbrains.signatureverifier.serialization.fileInfos

import com.jetbrains.signatureverifier.macho.CSMAGIC
import com.jetbrains.signatureverifier.macho.MachoConsts
import com.jetbrains.util.Jump
import com.jetbrains.util.Seek
import com.jetbrains.util.SeekOrigin
import kotlinx.serialization.Serializable
import java.nio.ByteBuffer
import java.nio.channels.SeekableByteChannel

@Serializable
data class MachoFileMetaInfo(
  var machoOffset: Long = 0L,
  var fileSize: Long = 0L,
  var isBe: Boolean = false,
  var headerMetaInfo: MachoHeaderMetaInfo = MachoHeaderMetaInfo(),
  val loadCommands: MutableList<LoadCommandInfo> = mutableListOf(),
  val codeSignatureInfo: CodeSignatureInfo = CodeSignatureInfo()
) : FileMetaInfo {


  override fun modifyFile(stream: SeekableByteChannel, signature: ByteArray) {
    stream.Seek(0, SeekOrigin.End)
    if (fileSize > stream.size()) {
      stream.write(ByteBuffer.wrap(ByteArray((fileSize - stream.size()).toInt() + 1)))
    }

    stream.Jump(machoOffset)

    stream.write(ByteBuffer.wrap(headerMetaInfo.toByteArray(isBe)))

    loadCommands.forEach {
      stream.Jump(it.offset + machoOffset)
      stream.write(ByteBuffer.wrap(it.toByteArray()))
    }

    stream.Jump(codeSignatureInfo.superBlobStart + machoOffset)
    stream.write(ByteBuffer.wrap(codeSignatureInfo.toByteArray()))
    codeSignatureInfo.blobs.forEach {
      stream.Jump(codeSignatureInfo.superBlobStart + machoOffset)
      stream.Seek(it.offset.toLong(), SeekOrigin.Current)

      if (it.magic == CSMAGIC.CMS_SIGNATURE) {
//        it.magic = MachoConsts.CSMAGIC_CMS_SIGNATURE.toUInt()
        it.length = signature.size + 8
        it.content = signature
      }
      if (it.type.toLong() == MachoConsts.CSSLOT_CODEDIRECTORY) {
        stream.write(ByteBuffer.wrap(it.content))
      } else {
        stream.write(ByteBuffer.wrap(it.toByteArray()))
      }
    }

  }
}